x/feeburner/keeper/keeper.go

```diff
@@ -71,6 +79,50 @@ func (k Keeper) GetTotalBurnedNeutronsAmount(ctx sdk.Context) types.TotalBurnedN
 	return totalBurnedNeutronsAmount
 }
 
+// BurnAndDistribute is an important part of tokenomics. It does few things:
+// 1. Burns NTRN fee coins distributed to consumertypes.ConsumerRedistributeName in ICS (https://github.com/cosmos/interchain-security/blob/v0.2.0/x/ccv/consumer/keeper/distribution.go#L17)
+// 2. Updates total amount of burned NTRN coins
+// 3. Sends non-NTRN fee tokens to treasury contract address
+// Panics if no `consumertypes.ConsumerRedistributeName` module found OR could not burn NTRN tokens
+func (k Keeper) BurnAndDistribute(ctx sdk.Context) error {
+	moduleAddr := k.accountKeeper.GetModuleAddress(consumertypes.ConsumerRedistributeName)
+	if moduleAddr == nil {
+		panic("ConsumerRedistributeName must have module address")
+	}
+
+	params := k.GetParams(ctx)
+	balances := k.bankKeeper.GetAllBalances(ctx, moduleAddr)
+	fundsForTreasury := make(sdk.Coins, 0, len(balances))
+
+	for _, balance := range balances {
+		if !balance.IsZero() {
+			if balance.Denom == params.NeutronDenom {
+				err := k.bankKeeper.BurnCoins(ctx, consumertypes.ConsumerRedistributeName, sdk.Coins{balance})
+				if err != nil {
+					panic(sdkerrors.Wrapf(err, "failed to burn NTRN tokens during fee processing"))
+				}
+
+				k.RecordBurnedFees(ctx, balance)
+			} else {
+				fundsForTreasury = append(fundsForTreasury, balance)
+			}
+		}
+	}
+
+	if len(fundsForTreasury) > 0 {
+		err := k.bankKeeper.SendCoins(
+			ctx,
+			moduleAddr, sdk.MustAccAddressFromBech32(params.TreasuryAddress),
+			fundsForTreasury,
+		)
+		if err != nil {
+			return fmt.Errorf("error sending funds to treasury for address=%s, tokens=%+v: %v", params.TreasuryAddress, fundsForTreasury, err)
```

Comment:
What is the specific reason of returning an error here? We panic everywhere in this method except this particular line.

Reply:
That's a bit out of scope). But:
I thought the reason was that it's okay to skip sending to treasury in this block, because we'll do it later anyway.
But now I think your point is valid actually, and this is not an okay behaviour.